src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,642 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace D2L.CodeStyle.Analyzers.Immutability;
+public sealed partial class ImmutabilityAnalyzer {
+
+	private static class ImmutableTypeParameterArgumentAnalysis {
+
+		public static void Register(
+			CompilationStartAnalysisContext context,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext
+		) {
+			// Type arguments of method return types
+			// Type arguments of method parameters
+			context.RegisterSymbolAction(
+				ctx => {
+					var symbol = (IMethodSymbol)ctx.Symbol;
+
+					// Ignore auto-implemened property methods
+					switch( symbol.MethodKind ) {
+						case MethodKind.PropertyGet:
+						case MethodKind.PropertySet:
+							return;
+					}
+
+					SyntaxNode getBaseSyntax() => symbol.DeclaringSyntaxReferences[ 0 ].GetSyntax( ctx.CancellationToken );
+
+					if( symbol.MethodKind != MethodKind.Constructor ) {
+						AnalyzeTypeRecursive(
+							ctx.ReportDiagnostic,
+							annotationsContext,
+							immutabilityContext,
+							symbol.ReturnType,
+							() => {
+								SyntaxNode syntax = getBaseSyntax();
+								return getBaseSyntax() switch {
+									MethodDeclarationSyntax methodDeclaration => methodDeclaration.ReturnType,
+									_ => syntax,
+								};
+							}
+						);
+					}
+
+					for( int i = 0; i < symbol.Parameters.Length; ++i ) {
+						IParameterSymbol parameter = symbol.Parameters[ i ];
+
+						AnalyzeTypeRecursive(
+							ctx.ReportDiagnostic,
+							annotationsContext,
+							immutabilityContext,
+							parameter.Type,
+							() => {
+								SyntaxNode syntax = getBaseSyntax();
+
+								return syntax switch {
+									MethodDeclarationSyntax methodDeclaration => methodDeclaration.ParameterList.Parameters[ i ].Type,
+									RecordDeclarationSyntax recordDeclaration => recordDeclaration.ParameterList!.Parameters[ i ].Type,

```

Comment:
This and SymbolKind.Parameter give duplicate diagnostics. IntelliSense shows the duplicates, but actually building de-dupes them into a single diagnostic.

Not sure how to skip one or the other without digging into syntax ahead of time.

Reply:
Bummer. You could emit them on CompilationEnd (do dedupe yourself) but that sucks too.